File: java/src/main/java/net/razorvine/serpent/ComplexNumber.java
/**
 * Serpent, a Python literal expression serializer/deserializer
 * (a.k.a. Python's ast.literal_eval in Java)
 * Software license: "MIT software license". See http://opensource.org/licenses/MIT
 * @author Irmen de Jong (irmen@razorvine.net)
 */

package net.razorvine.serpent;

import java.io.Serializable;

/**
 * A complex number.
 */
public class ComplexNumber implements Serializable {
	
	private static final long serialVersionUID = 5396759273405612137L;
	
	public double real;
	public double imaginary;
	
	public ComplexNumber(double r, double i) {
		real=r;
		imaginary=i;
	}

	@Override
	public String toString()
	{
		StringBuilder sb=new StringBuilder();
		sb.append(real);
		if(imaginary>0)
			sb.append('+');
		return sb.append(imaginary).append('i').toString();
	}
	
	public double Magnitude() {
		return Math.sqrt(real * real + imaginary * imaginary);
	}

	public void add(ComplexNumber other)
	{
		real += other.real;
		imaginary += other.imaginary;
	}

	public void subtract(ComplexNumber other)
	{
		real -= other.real;
		imaginary -= other.imaginary;
	}

	public void multiply(ComplexNumber other)
	{
		double new_real = real * other.real - imaginary * other.imaginary;
		double new_imaginary = real * other.imaginary + imaginary * other.real;
		real = new_real;
		imaginary = new_imaginary;
	}

	public void divide(ComplexNumber other)
	{
		double new_real = (real * other.real + imaginary * other.imaginary) / (other.real * other.real + other.imaginary * other.imaginary);
		double new_imaginary = (imaginary * other.real - real * other.imaginary) / (other.real * other.real + other.imaginary * other.imaginary);
		real = new_real;
		imaginary = new_imaginary;
	}
	
	@Override
	public boolean equals(Object obj)
	{
		if(!(obj instanceof ComplexNumber))
		{
			return false;
		}
		ComplexNumber other = (ComplexNumber) obj;
		return real==other.real && imaginary==other.imaginary;
	}
	
	@Override
	public int hashCode()
	{
		Double r = Double.valueOf(real);
		Double i = Double.valueOf(imaginary);
		return r.hashCode() ^ i.hashCode();
	}
}

File: dotnet/Serpent/Razorvine.Serpent/ComplexNumber.cs
using System;
using System.Text;
// ReSharper disable UnusedMember.Global

namespace Razorvine.Serpent
{

/// <summary>
/// A Complex Number class. 
/// </summary>
public class ComplexNumber {
	
	public double Real {get; }
	public double Imaginary {get; }
	
	public ComplexNumber(double r, double i) {
		Real=r;
		Imaginary=i;
	}

	public override string ToString()
	{
		var sb=new StringBuilder();
		sb.Append(Real);
		if(Imaginary>0)
			sb.Append('+');
		return sb.Append(Imaginary).Append('i').ToString();
	}

	public double Magnitude() {
		return Math.Sqrt(Real * Real + Imaginary * Imaginary);
	}

	public static ComplexNumber operator +(ComplexNumber c1, ComplexNumber c2) {
		return new ComplexNumber(c1.Real + c2.Real, c1.Imaginary + c2.Imaginary);
	}


	public static ComplexNumber operator -(ComplexNumber c1, ComplexNumber c2) {
		return new ComplexNumber(c1.Real - c2.Real, c1.Imaginary - c2.Imaginary);
	}

	public static ComplexNumber operator *(ComplexNumber c1, ComplexNumber c2) {
		return new ComplexNumber(c1.Real * c2.Real - c1.Imaginary * c2.Imaginary, c1.Real * c2.Imaginary + c1.Imaginary * c2.Real);
	}

	public static ComplexNumber operator /(ComplexNumber c1, ComplexNumber c2) {
		return new ComplexNumber((c1.Real * c2.Real + c1.Imaginary * c2.Imaginary) / (c2.Real * c2.Real + c2.Imaginary * c2.Imaginary), (c1.Imaginary * c2.Real - c1.Real * c2.Imaginary)
				/ (c2.Real * c2.Real + c2.Imaginary * c2.Imaginary));
	}

	#region Equals and GetHashCode implementation
	public override bool Equals(object obj)
	{
		if(obj is not ComplexNumber other)
			return false;
		// ReSharper disable CompareOfFloatsByEqualityOperator
		return Real==other.Real && Imaginary==other.Imaginary;
	}
	
	public override int GetHashCode()
	{
		return Real.GetHashCode() ^ Imaginary.GetHashCode();
	}
	
	public static bool operator ==(ComplexNumber lhs, ComplexNumber rhs)
	{
		if (ReferenceEquals(lhs, rhs))
			return true;
		if (ReferenceEquals(lhs, null) || ReferenceEquals(rhs, null))
			return false;
		return lhs.Equals(rhs);
	}
	
	public static bool operator !=(ComplexNumber lhs, ComplexNumber rhs)
	{
		return !(lhs == rhs);
	}
	#endregion
}

}